fix  VAE loading issue in train_dreambooth

bssrdf opened the pull request:
# What does this PR do?
This PR is similar to https://github.com/huggingface/diffusers/pull/3462 but fixes the same issue in ```train_dreambooth.py```.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes https://github.com/huggingface/diffusers/issues/7619


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@sayakpaul @standardAI 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @yiyixuxu 
- Pipelines:  @sayakpaul @yiyixuxu @DN6
- Training examples: @sayakpaul 
- Docs: @stevhliu and @sayakpaul
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @sayakpaul @yiyixuxu @DN6

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

sayakpaul commented:
We need to make sure the code quality tests pass.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_7632). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

standardAI commented:
@bssrdf could you run [`make style && make quality`](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md#how-to-open-a-pr:~:text=%F0%9F%A7%A8%20Diffusers%20relies%20on,%24%20make%20quality)?

bssrdf commented:
> @bssrdf could you run [`make style && make quality`](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md#how-to-open-a-pr:~:text=%F0%9F%A7%A8%20Diffusers%20relies%20on,%24%20make%20quality)?

Done!

sayakpaul commented:
It still doesn't pass :/

bssrdf commented:
> It still doesn't pass :/

I don't know what's going on. Any way, I rerun ```make style``` and ```make quality``` and here are the outputs. Nothing stands out.
```
$ make style
ruff check examples scripts src tests utils benchmarks setup.py --fix
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'ignore' -> 'lint.ignore'
  - 'select' -> 'lint.select'
  - 'isort' -> 'lint.isort'
  - 'per-file-ignores' -> 'lint.per-file-ignores'
All checks passed!
ruff format examples scripts src tests utils benchmarks setup.py
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'ignore' -> 'lint.ignore'
  - 'select' -> 'lint.select'
  - 'isort' -> 'lint.isort'
  - 'per-file-ignores' -> 'lint.per-file-ignores'
876 files left unchanged
doc-builder style src/diffusers docs/source --max_len 119
make autogenerate_code
make[1]: Entering directory '/home/isodden/temp/diffusers'
running deps_table_update
updating src/diffusers/dependency_versions_table.py
make[1]: Leaving directory '/home/isodden/temp/diffusers'
make extra_style_checks
make[1]: Entering directory '/home/isodden/temp/diffusers'
python utils/custom_init_isort.py
python utils/check_doc_toc.py --fix_and_overwrite
make[1]: Leaving directory '/home/isodden/temp/diffusers'
```

```
$make quality
ruff check examples scripts src tests utils benchmarks setup.py
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'ignore' -> 'lint.ignore'
  - 'select' -> 'lint.select'
  - 'isort' -> 'lint.isort'
  - 'per-file-ignores' -> 'lint.per-file-ignores'
All checks passed!
ruff format --check examples scripts src tests utils benchmarks setup.py
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'ignore' -> 'lint.ignore'
  - 'select' -> 'lint.select'
  - 'isort' -> 'lint.isort'
  - 'per-file-ignores' -> 'lint.per-file-ignores'
876 files already formatted
doc-builder style src/diffusers docs/source --max_len 119 --check_only
python utils/check_doc_toc.py
```

```
$git log
commit ee9b19e53e4ff1b14a001fadc140f19c47959bb1 (HEAD -> fix-train-dreambooth-vae-not-loading, origin/fix-train-dreambooth-vae-not-loading)
Merge: ac68fe51 279de3c3
Author: Sayak Paul <spsayakpaul@gmail.com>
Date:   Fri Apr 12 08:33:05 2024 +0530

    Merge branch 'main' into fix-train-dreambooth-vae-not-loading

commit 279de3c3ffedcb1394518a8f1c950fa30f272390
Author: Sai-Suraj-27 <sai.suraj.27.729@gmail.com>
Date:   Fri Apr 12 01:13:01 2024 +0530

    fix: Replaced deprecated `logger.warn` with `logger.warning` (#7643)

    Fixed deprecated logger.warn with logger.warning.

commit 8e14535708f6af0794148150f5c073c4723dbbae
Author: Yiqin Zhao <yiqinzhao@outlook.com>
Date:   Thu Apr 11 15:08:42 2024 -0400

    Fixed YAML loading. (#7579)

commit 0bee4d336b925b6064eee156f5a155e3ca3b30ab
Author: dg845 <58458699+dg845@users.noreply.github.com>
Date:   Thu Apr 11 10:52:12 2024 -0700

    LCM Distill Scripts Fix Bug when Initializing Target U-Net (#6848)

    * Initialize target_unet from unet rather than teacher_unet so that we correctly add time_embedding.cond_proj if necessary.

    * Use UNet2DConditionModel.from_config to initialize target_unet from unet's config.

    ---------

    Co-authored-by: Sayak Paul <spsayakpaul@gmail.com>

commit ac68fe5151a7e1bba83167bd07b65aded716174b
Author: bssrdf <bssrdf@gmail.com>
Date:   Thu Apr 11 09:22:17 2024 -0400

    reformat code with make style && make quality

```

standardAI commented:
Could you reverse the last changes `make style && make quality` done, and [downgrade your ruff](https://github.com/huggingface/diffusers/blob/2523390c269a37209fbb96359d494369aff89361/setup.py#L124), then run then command again?

bssrdf commented:
I downgraded ruff and reran the "make style && make quality". Let's give it another push. Yeah, ruff is now on 0.3.5. diffusers uses 0.1.5 which is pretty old.

standardAI reviewed (approved):
It seems OK now. Thanks for fixing this!

bghira commented on examples/dreambooth/train_dreambooth.py:
```diff
@@ -933,11 +923,13 @@ def main(args):
 
-    if model_has_vae(args):
+    try:
         vae = AutoencoderKL.from_pretrained(
             args.pretrained_model_name_or_path, subfolder="vae", revision=args.revision, variant=args.variant
         )
-    else:
+    except OSError:
+        # IF does not have a VAE so let's just set it to None
+        # We don't have to error out here
         vae = None
```
no, we can't blindly just set None for the VAE if it fails to load. assuming it's DeepFloyd isn't great, the reason that check was appropriate before is because model_has_vae would check for an actual VAE config in the repo.

bssrdf replied:
that check ```model_has_vae``` is broken, hence this PR and https://github.com/huggingface/diffusers/pull/3462. I am not the first one who encountered this problem. This PR will make train_dream work for users until the check is fixed.

bghira replied:
in fact, it needs to be fixed in both scripts.

bssrdf replied:
I figured out the problem, which is specific only on Windows OS. Please review the updated PR with ```model_has_vae``` check augmented.

bghira replied:
thanks for being patient with me and looking into that deeper. that is a good fix to find.

bghira commented:
i understand that the other script does this, but that change is from May, 2023. if anything we should fix the method used by this script, and enhance the other scripts to make use of the correct method.

i don't think this should be merged, it's regressing the code quality.

bghira commented on examples/dreambooth/train_dreambooth.py:
```diff
@@ -757,6 +757,16 @@ def __getitem__(self, index):
+def model_has_vae(args):
+    config_file_name = os.path.join("vae", AutoencoderKL.config_name)
+    if platform.system() == "Windows":
+        config_file_name = config_file_name.replace('\\', '/')
```
ohhhhhhh!!! i never use Windows, so this explains a lot.

the os.path.join is actually something we can remove then! it can be `f"vae/{AutoencoderKL.config_name}"` on line 761 and then the Windows check can go away.

I guess the Git info response will have Linux-like paths in it always.

bssrdf replied:
I removed os.path.join for config_file_name and used '/' for concatenation. It is OS independent now. Thanks for suggestions.

bghira commented on examples/dreambooth/train_dreambooth.py:
```diff
@@ -757,6 +757,16 @@ def __getitem__(self, index):
+        config_file_name = config_file_name.replace('\\', '/')
+    if os.path.isdir(args.pretrained_model_name_or_path):
+        config_file_name = os.path.join(args.pretrained_model_name_or_path, config_file_name)
+        return os.path.isfile(config_file_name)
```
these should remain the same using `os.*`

bssrdf replied:
These stay the same

sayakpaul commented on examples/dreambooth/train_dreambooth.py:
```diff
@@ -757,17 +757,15 @@ def __getitem__(self, index):
-
 def model_has_vae(args):
-    config_file_name = os.path.join("vae", AutoencoderKL.config_name)
+    config_file_name = f"vae/{AutoencoderKL.config_name}"
```
Could we make use of `config_file_name = Path("vae", AutoencoderKL.config_name).as_posix()` here?

bssrdf replied:
@sayakpaul, this will also work. I can change to using Path and as_posix if we all agree it is proper to do.

sayakpaul replied:
I think that will be better to promote generality.

bghira replied:
i didn't even know that exists. learning things!

bssrdf replied:
Done! Thanks for the tips. I too learned a thing.

sayakpaul reviewed (commented):
Thanks for finding this. Left a single comment.

yiyixuxu commented:
@sayakpaul can we merge this now?

sayakpaul commented:
Sorry for the delay. Once this CI is green, will merge.
